module/zfs/dmu_send.c

```diff
@@ -1598,18 +1603,30 @@ restore_freeobjects(struct restorearg *ra, objset_t *os,
 	if (drrfo->drr_firstobj + drrfo->drr_numobjs < drrfo->drr_firstobj)
 		return (SET_ERROR(EINVAL));
 
-	for (obj = drrfo->drr_firstobj;
-	    obj < drrfo->drr_firstobj + drrfo->drr_numobjs;
-	    (void) dmu_object_next(os, &obj, FALSE, 0)) {
+	obj = drrfo->drr_firstobj;
+
+	if (obj == 0)
+		obj++;
+
+	while (obj < drrfo->drr_firstobj + drrfo->drr_numobjs) {
```

Comment:
Performance of this will be very bad for certain streams, because you are not using dmu_object_next() to advance to the next allocated object.